### ISSUE TITLE
[QA/REVIEW] DPR retrieval 학습 관련 질문입니다.

### ISSUE THREAD

**Issue by devholic357** [VumBleBot/odqa_baseline_code#112]
##  📸  코드

- 코드나 코드에 대한 스크린샷을 첨부해주세요.

---
![vumblebot_prepare](https://user-images.githubusercontent.com/88775006/137060058-0983da5d-c412-4be7-9ee1-009d2a925fcf.PNG)
 

## 🤔  궁금한 점

- 궁금한 점, 리뷰할 점을 상세히 기술해주세요. 

안녕하세요, 좋은 repository 공유해주셔서 감사합니다. 
retrieval 의 DPRBERT embedding을 처음부터 다시 학습하려고 하는데, 궁금한점이 있어서 문의드립니다.
README 의 내용을 참고해서, "retriever_name": "DPRBERT", "retrain": true, "dense_train_dataset": "train_dataset", 등으로 설정하고 30 epoch을 돌렸는데, loss 는 떨어지는데 retrieve된 결과가 좀 이상한것 같습니다.

DPRBERT 학습 시, BaseTrainMixin 의 _train() 함수로 학습을 시작하는데, 혹시 제가 뭔가 잘못 진행하고 있는걸까요?
가능하다면, DPRBERT 학습에 필요한 전략 파일 설정을 받을 수 있을까요?
감사합니다!

아래는 제가 사용했던 전략 파일 내용입니다. 
{
    "alias": "EXAMPLES DPRBERT",
    "model": {
        "retriever_name": "DPRBERT",
        "tokenizer_name": ""
    },
    "data": {
        "dataset_name": "squad_kor_v1",
        "sub_datasets": ""
    },
    "train": {
        "save_total_limit": 2,
        "save_steps": 100,
        "logging_steps": 100,
        "overwrite_output_dir": true,
        "do_train": true,
        "do_eval": true,
        "do_predict": false,
        "report_to": ["wandb"]
    },
    "retriever": {
        "retrain": true,
        "dense_train_dataset": "train_dataset",
        "topk": 10,
        "learning_rate": 3e-5,
        "per_device_train_batch_size": 4,
        "per_device_eval_batch_size": 4,
        "num_train_epochs": 30,
        "weight_decay": 0.01
    }
}

## 📚  레퍼런스 첨부

- 의문 해결에 도움을 주는 참고 문헌이 있다면 기록해주세요!
- 책, 블로그, 유튜브, 논문 등

**Comment by ggm1207** [comment]
@devholic357

안녕하세요, 답변이 늦었습니다. 질문자께서 이미 해결하신 듯 하지만 같은 의문을 가진 다른 분들께도 유용한 답변이 될 것 같아서 남겨봅니다.

우선 성능이 낮게 나오는 이유는 지금 구현된 DPRRetrieval을 학습시키는 것이 하이퍼파라미터 세팅에 매우 의존적이라 어렵기 때문입니다. 더 나은 구현체를 사용한다면 성능이 잘 나올 수도 있습니다.

> 저희 또한 이 Repository를 작성하면서 약 100번 넘게 학습을 돌려봤는데 한 번을 제외하고 대부분 TOP 10에서 20%(Accuracy)의 성능이 나왔습니다. 가장 높았던 성능이 TOP 10에서 50%(Accuracy) 였습니다.

두 번째 이유는 DPRRetrieval 클래스가 딥러닝 방식 문서 검색의 초창기 모델이라 SparseRetrieval에 비해 기본적으로 성능이 좋지 않습니다. 질문의 답변은 첫 번째 이유로도 충분하지만 아시면 좋을 것 같아서 덧붙여서 남겨봅니다.

- [Retrieving and Reading : A Comprehensive Survey on Open-domain Question Answering](https://arxiv.org/pdf/2101.00774.pdf)

위 논문에서 딥러닝 방식의 Retriever를 설명하는데 크게 3가지의 종류가 있습니다.

1. **Representation-based Retriever**
    - DPRRetrieval가 여기에 속합니다.
2. **Interaction-based Retriever**
3. **Representation-interaction Retriever**
    - 레포에 구현되어 있는 ColBert가 여기에 속합니다.

아래로 갈수록 더 최신의 방법이며 논문에서는 SparseRetriever의 성능을 DenseRetriever가 뛰어 넘은지는 그리 오래 기간이 지나지 않았다고 합니다.

> 주관적인 생각으로는 DPRRetrieval 자체가 SparseRetrieval을 뛰어 넘기 전의 모델이라고 생각을 합니다.

예전에 공부했던 내용이라 틀린 부분이 있을 수 있습니다.
틀린 부분이 있다면 아래에 추가로 댓글 달아주시면 감사하겠습니다!
